astropy/io/ascii/tdat.py

```diff
@@ -0,0 +1,815 @@
+        """
+        # if we have  delimiter from the header, use it.
+        field_delimiter = getattr(self.header, "delimiter", self.splitter.delimiter)
+        self.splitter.delimiter = field_delimiter
+        record_delimiter = getattr(
+            self.header, "record_delimiter", self.splitter.record_delimiter
+        )
+        allowed_delimiters = list(map(chr, range(1, 128)))
+        if record_delimiter is not None:
+            if record_delimiter not in allowed_delimiters:
+                raise TdatFormatError(
+                    f"Unsupported record delimiter: {record_delimiter}."
+                )
+            if record_delimiter == "|":
+                raise TdatFormatError("'|' character reserved for field delimiter.")
+        for c in field_delimiter:
+            if c not in allowed_delimiters:
+                raise TdatFormatError(f"Unsupported field delimiter: {c}.")
+        nlines = len(self.header._line_fields)
+        return self.splitter(self.data_lines, field_delimiter, record_delimiter, nlines)
+
+    def write(self, lines):
+        """
+        WRITE: Override the default write to include <DATA> and <END> lines.
+        """
+        lines.append("<DATA>")
+        super().write(lines)
+        lines.append("<END>")
+
+
+class TdatOutputter(core.TableOutputter):
+    """
+    Output the table as an astropy.table.Table object.
+    """
+
+    def __call__(self, cols, meta):
+        """
+        READ: Override the default outputter.
+        TDAT files may (optionally) specify which field lines should be used as
+        the primary index and secondary indices Astropy tables support adding
+        indices after creation. This overwrite adds labeled indices on read.
+        """
+        # Sets col.data to numpy array and col.type to io.ascii Type class (e.g.
+        # FloatType) for each col.
+        self._convert_vals(cols)
+
+        t_cols = [
+            np.ma.MaskedArray(x.data, mask=x.mask)
+            if hasattr(x, "mask") and np.any(x.mask)
+            else x.data
+            for x in cols
+        ]
+        out = core.Table(t_cols, names=[x.name for x in cols], meta=meta["table"])
+        indices = []
+        for col, out_col in zip(cols, out.columns.values()):
+            for attr in ("format", "unit", "description"):
+                if hasattr(col, attr):
+                    setattr(out_col, attr, getattr(col, attr))
+            if hasattr(col, "meta"):
+                out_col.meta.update(col.meta)
+                if "index" in col.meta:
+                    if col.meta["index"] == "key":
+                        indices.insert(0, col.name)
+                    else:
+                        indices.append(col.name)
+        # Add indices, if specified
+        if len(indices) > 0:
+            for name in indices:
+                out.add_index(name)
+        return out
+
+
+class Tdat(core.BaseReader):
+    """TDAT format
```

Comment:
Overall for the docstring here, you should build the docs from source and make sure they look good before pushing docs commits to GitHub. Sphinx/RST is really fiddly and I never get it right the first time, so building locally prevents burning CI hours.

https://docs.astropy.org/en/latest/development/docguide.html#building-the-documentation-from-source